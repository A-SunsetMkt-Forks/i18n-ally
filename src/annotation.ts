import * as vscode from 'vscode'
import { debounce } from 'lodash'
import { KEY_REG } from './utils/KeyDetector'
import i18nFiles from './utils/i18nFiles'
import Common from './utils/Common'

const textEditorDecorationType = vscode.window.createTextEditorDecorationType(
  {}
)

function annotation (ctx: vscode.ExtensionContext) {
  function update () {
    const activeTextEditor = vscode.window.activeTextEditor
    if (!activeTextEditor)
      return

    const { document } = activeTextEditor
    const text = document.getText()
    const decorations = []

    // 从文本里遍历生成中文注释
    let match = null
    while (match = KEY_REG.exec(text)) {
      const index = match.index
      const matchKey = match[0]
      const key = matchKey.replace(new RegExp(KEY_REG), '$1')
      const trans = i18nFiles.getTransByKey(document.fileName, key)

      if (!trans)
        return

      const text = (trans.find(t => t.lng === Common.displayLanguage) || { data: '' }).data || ''
      const decoration = {
        range: new vscode.Range(
          document.positionAt(index),
          document.positionAt(index + matchKey.length + 1)
        ),
        renderOptions: {
          after: {
            color: 'rgba(153, 153, 153, .7)',
            contentText: `◽️${text || '""'}`,
            fontWeight: 'normal',
            fontStyle: 'normal',
          },
        },
      }
      decorations.push(decoration)

      activeTextEditor.setDecorations(textEditorDecorationType, decorations)
    }
  }

  const debounceUpdate = debounce(update, 500);

  [
    vscode.window.onDidChangeActiveTextEditor,
    vscode.workspace.onDidChangeTextDocument,
  ].forEach((onChange: any) => {
    onChange(debounceUpdate, null, ctx.subscriptions)
  })

  Common.localesPaths.forEach(i18nPath => {
    const i18nDirWatcher = vscode.workspace.createFileSystemWatcher(
      `${i18nPath}/**`
    )

    i18nDirWatcher.onDidChange(debounceUpdate)
    i18nDirWatcher.onDidCreate(debounceUpdate)
    i18nDirWatcher.onDidDelete(debounceUpdate)
    ctx.subscriptions.push(i18nDirWatcher)
  })

  update()
}

export default annotation
