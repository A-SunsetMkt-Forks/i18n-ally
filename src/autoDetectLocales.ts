import * as vscode from 'vscode'
import * as fg from 'fast-glob'
import * as path from 'path'
import Common from './utils/Common'

class AutoInit {
  ctx: vscode.ExtensionContext

  constructor (ctx: vscode.ExtensionContext) {
    this.ctx = ctx
  }

  async init () {
    const i18nPaths = Common.i18nPaths
    if (i18nPaths.length)
      return

    this.autoSet()
  }

  async autoSet () {
    const rootPath = vscode.workspace.rootPath
    const pattern = [`${rootPath}/**/(locales|locale)`]
    let result: string[] = await fg(pattern, {
      ignore: ['**/node_modules'],
      onlyDirectories: true,
    })

    result = result.map(r => path.relative(rootPath, r))

    Common.updateI18nPaths(result)

    await vscode.window.showInformationMessage(
      `VueI18n locales path auto set to ${result.join(';').toString()}`,
    )
  }
}

export default (ctx: vscode.ExtensionContext) => {
  const autoInit = new AutoInit(ctx)
  autoInit.init()

  return vscode.commands.registerCommand('extension.vue-i18n-ally.auto-detect-locales', () => {
    autoInit.autoSet()
  })
}
