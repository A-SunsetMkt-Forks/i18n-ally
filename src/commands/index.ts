import { flatten } from 'lodash'
import { ExtensionModule } from '../modules'
import configLocales from './configLocales'
import configLanguages from './configLanguages'
import keyManipulations from './keyManipulations'
import extractText from './extractText'
import help from './help'
import refreshUsageReport from './refreshUsageReport'
import editor from './editor'

const m: ExtensionModule = (ctx) => {
  return flatten([
    configLocales(ctx),
    configLanguages(ctx),
    keyManipulations(ctx),
    extractText(ctx),
    help(ctx),
    refreshUsageReport(ctx),
    editor(ctx),
  ])
}

export default m
