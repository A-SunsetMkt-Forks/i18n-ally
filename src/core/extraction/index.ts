export * from './detect'
export * from './rules'
