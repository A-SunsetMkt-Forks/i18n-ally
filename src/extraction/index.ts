export * from './shouldExtract'
export * from './rules'
export * from './parsers'
