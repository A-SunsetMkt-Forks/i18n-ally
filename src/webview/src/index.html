<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf8" />
    <meta name="viewport" content="width=device-width" />
    <title>i18n Ally Editor</title>
    <style>
      html {
        background-color: var(--vscode-editor-background);
        color: var(--vscode-editor-foreground);
        font-family: var(--vscode-font-family);
        font-weight: var(--vscode-font-weight);
        font-size: var(--vscode-font-size);
      }
    </style>
  </head>

  <body>
    <div id="app"></div>
    <script src="./index.js"></script>
  </body>
</html>
