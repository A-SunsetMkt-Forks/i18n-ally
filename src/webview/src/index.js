/* eslint-disable no-undef */
import Vue from 'vue'
import Vuex from 'vuex'
import 'vue-material-design-icons/styles.css'
import VueI18n from 'vue-i18n'
import VCheck from 'vue-material-design-icons/Check.vue'
import VPlusMinus from 'vue-material-design-icons/PlusMinus.vue'
import VCommentOutline from 'vue-material-design-icons/CommentOutline.vue'
import VEarth from 'vue-material-design-icons/Earth.vue'
import VCommentEditOutline from 'vue-material-design-icons/CommentEditOutline.vue'
import VCommentQuestionOutline from 'vue-material-design-icons/CommentQuestionOutline.vue'
import VCheckboxMarkedOutline from 'vue-material-design-icons/CheckboxMarkedOutline.vue'
import VPencilOff from 'vue-material-design-icons/PencilOff.vue'
import VPencil from 'vue-material-design-icons/Pencil.vue'
import VCheckAll from 'vue-material-design-icons/CheckAll.vue'
import VMenu from 'vue-material-design-icons/Menu.vue'
import VChevronLeft from 'vue-material-design-icons/ChevronLeft.vue'
import VChevronRight from 'vue-material-design-icons/ChevronRight.vue'
import VDeleteEmptyOutline from 'vue-material-design-icons/DeleteEmptyOutline.vue'
import VFormatQuoteOpen from 'vue-material-design-icons/FormatQuoteOpen.vue'
import { api, mode } from './api'
import App from './App.vue'

Vue.component('VCheck', VCheck)
Vue.component('VPlusMinus', VPlusMinus)
Vue.component('VCommentOutline', VCommentOutline)
Vue.component('VEarth', VEarth)
Vue.component('VCommentEditOutline', VCommentEditOutline)
Vue.component('VCommentQuestionOutline', VCommentQuestionOutline)
Vue.component('VCheckboxMarkedOutline', VCheckboxMarkedOutline)
Vue.component('VPencilOff', VPencilOff)
Vue.component('VPencil', VPencil)
Vue.component('VCheckAll', VCheckAll)
Vue.component('VDeleteEmptyOutline', VDeleteEmptyOutline)
Vue.component('VFormatQuoteOpen', VFormatQuoteOpen)
Vue.component('VMenu', VMenu)
Vue.component('VChevronLeft', VChevronLeft)
Vue.component('VChevronRight', VChevronRight)

Vue.use(Vuex)
Vue.use(VueI18n)

const locale = 'en'
const i18n = new VueI18n({
  locale,
  messages: {},
})

function initTheme() {
  if (mode === 'webview')
    return
  const theme = {
    foreground: 'white',
    'editor-foreground': 'white',
    background: '#222',
    'editor-background': '#222',
    'font-family': '-apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif',
    'editor-font-family': 'monospace',
  }
  const css = []
  for (const key of Object.keys(theme))
    css.push(`--vscode-${key}:${theme[key]}`)

  const style = document.createElement('style')
  style.innerHTML = `:root{${css.join(';')}}`
  document.head.appendChild(style)
}

initTheme()

const store = new Vuex.Store({
  state: () => {
    return Object.assign({
      mode,
      ready: false,
      config: {
        debug: false,
        sourceLanguage: 'en',
        displayLanguage: 'en',
        enabledFrameworks: [],
        ignoredLocales: [],
        extensionRoot: '',
        flags: [],
        locales: [],
      },
      context: {},
      i18n: {},
      route: 'welcome',
      routeData: {},
    },
    { ready: false })
  },
  mutations: {
    config(state, data) {
      state.config = data
    },
    i18n(state, data) {
      state.i18n = data
      i18n.setLocaleMessage(locale, data)
    },
    route(state, { route, data }) {
      state.routeData = data
      state.route = route
    },
    context(state, context) {
      state.context = context
    },
    ready(state) {
      state.ready = true
    },
  },
})

api.registerListener((message) => {
  switch (message.type) {
    case 'ready':
      store.commit('ready')
      break
    case 'config':
      store.commit('config', message.data)
      break
    case 'route':
      store.commit('route', message)
      break
    case 'i18n':
      store.commit('i18n', message.data)
      break
    case 'context':
      store.commit('context', message.data)
  }
})

// @ts-ignore
window.app = new Vue({
  store,
  i18n,
  render: createElement => createElement(App),
}).$mount('#app')

api.postMessage({ type: 'ready' })
