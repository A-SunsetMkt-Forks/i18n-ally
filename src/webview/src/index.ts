/* eslint-disable no-undef */
// @ts-ignore
import Vue from 'vue'
// @ts-ignore
import Vuex from 'vuex'
// @ts-ignore
import VueI18n from 'vue-i18n'
import './icons'
import { api } from './api/index'
import 'vue-material-design-icons/styles.css'
// @ts-ignore
import App from './components/App.vue'

Vue.use(Vuex)
Vue.use(VueI18n)

const locale = 'en'
const i18n = new VueI18n({
  locale,
  messages: {},
})

const store = new Vuex.Store({
  state: {
    mode: api.mode,
    ready: false,
    config: {
      debug: false,
      sourceLanguage: 'en',
      displayLanguage: 'en',
      enabledFrameworks: [],
      ignoredLocales: [],
      extensionRoot: '',
      flags: [],
      locales: [],
    },
    client: {
      ready: false,
    },
    sidebar: false,
    context: {
      standalone: false,
      keys: [],
      filepath: '',
    },
    i18n: {},
    route: 'welcome',
    routeData: {
      index: 0,
    },
    pendingChanges: [],
  },
  mutations: {
    config(state, data) {
      state.config = data
    },
    i18n(state, data) {
      state.i18n = data
      i18n.setLocaleMessage(locale, data)
    },
    route(state, { route, data }) {
      state.routeData = data
      state.route = route
    },
    context(state, context) {
      state.context = context
    },
    ready(state) {
      state.ready = true
    },
    clientReady(state, v) {
      state.client.ready = v
    },
    set(state, { key, value }) {
      Vue.set(state, key.split('.'), value)
    },
  },
})

api.server.registerListener((message) => {
  switch (message.type) {
    case 'ready':
      store.commit('ready')
      break
    case 'config':
      store.commit('config', message.data)
      break
    case 'route':
      store.commit('route', message)
      break
    case 'i18n':
      store.commit('i18n', message.data)
      break
    case 'context':
      store.commit('context', message.data)
  }
})

if (api.devtools) {
  api.devtools.registerListener((message) => {
    switch (message.type) {
      case 'devtools.ready':
        store.commit('clientReady', true)
        break
      case 'devtools.close':
        store.commit('clientReady', false)
        break
      case 'context':
        store.commit('context', message.data)
    }
  })
}

// @ts-ignore
window.app = new Vue({
  store,
  i18n,
  render: createElement => createElement(App),
}).$mount('#app')

api.server.postMessage({ type: 'ready' })
