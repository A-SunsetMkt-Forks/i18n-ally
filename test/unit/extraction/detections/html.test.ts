import { expect } from 'chai'
import { extractionsParsers } from '../../../../src/extraction'

const html = extractionsParsers.html

describe('detections', () => {
  it('html', () => {
    expect(html.detect('<div placeholder="hello"></div>')).to.matchSnapshot()

    expect(html.detect(`
<textarea
  class="form-control"
  rows="8"
  v-model="article.body"
  placeholder="Write your article (in markdown)"
  ></textarea>
`)).to.matchSnapshot()

    expect(html.detect('<img src="https://example.com/img.svg" :title="image" alt="img.svg">')).to.matchSnapshot()
    expect(html.detect('<img src="{{ img_url }}" title="This is a title" alt="This is an alt.">')).to.matchSnapshot()
  })

  it('vue-favored', () => {
    expect(html.detect(
      // eslint-disable-next-line no-template-curly-in-string
      '<input v-bind:placeholder="\'Placeholder \' + input_name" :title="`${input_name} title`" name="email">'),
    ).to.matchSnapshot()
  })
})
