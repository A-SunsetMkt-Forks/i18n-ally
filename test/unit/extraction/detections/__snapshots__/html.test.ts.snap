// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`detections html 1`] = `
Array [
  Object {
    "end": 23,
    "fullEnd": 24,
    "fullStart": 5,
    "fullText": "placeholder=\\"hello\\"",
    "isDynamic": false,
    "start": 18,
    "text": "hello",
  },
]
`;

exports[`detections html 2`] = `
Array [
  Object {
    "end": 117,
    "fullEnd": 118,
    "fullStart": 72,
    "fullText": "placeholder=\\"Write your article (in markdown)\\"",
    "isDynamic": false,
    "start": 85,
    "text": "Write your article (in markdown)",
  },
]
`;

exports[`detections html 3`] = `Array []`;

exports[`detections html 4`] = `
Array [
  Object {
    "end": 47,
    "fullEnd": 48,
    "fullStart": 25,
    "fullText": "title=\\"This is a title\\"",
    "isDynamic": false,
    "start": 32,
    "text": "This is a title",
  },
  Object {
    "end": 69,
    "fullEnd": 70,
    "fullStart": 49,
    "fullText": "alt=\\"This is an alt.\\"",
    "isDynamic": false,
    "start": 54,
    "text": "This is an alt.",
  },
]
`;

exports[`detections vue-favored 1`] = `
Array [
  Object {
    "end": 54,
    "fullEnd": 55,
    "fullStart": 7,
    "fullText": "v-bind:placeholder=\\"'Placeholder ' + input_name\\"",
    "isDynamic": true,
    "start": 27,
    "text": "'Placeholder ' + input_name",
  },
]
`;
